solr/contrib/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java

```diff
@@ -210,50 +216,59 @@ public void setContext(ResultContext context) {
       }
       
       // Setup LTRScoringQuery
-      scoringQuery = SolrQueryRequestContextUtils.getScoringQuery(req);
-      docsWereNotReranked = (scoringQuery == null);
-      String featureStoreName = SolrQueryRequestContextUtils.getFvStoreName(req);
-      if (docsWereNotReranked || (featureStoreName != null && (!featureStoreName.equals(scoringQuery.getScoringModel().getFeatureStoreName())))) {
-        // if store is set in the transformer we should overwrite the logger
-
-        final ManagedFeatureStore fr = ManagedFeatureStore.getManagedFeatureStore(req.getCore());
-
-        final FeatureStore store = fr.getFeatureStore(featureStoreName);
-        featureStoreName = store.getName(); // if featureStoreName was null before this gets actual name
-
-        try {
-          final LoggingModel lm = new LoggingModel(loggingModelName,
-              featureStoreName, store.getFeatures());
-
-          scoringQuery = new LTRScoringQuery(lm,
-              LTRQParserPlugin.extractEFIParams(localparams),
-              true,
-              threadManager); // request feature weights to be created for all features
-
-        }catch (final Exception e) {
-          throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
-              "retrieving the feature store "+featureStoreName, e);
-        }
-      }
+      rerankingQueries = SolrQueryRequestContextUtils.getScoringQueries(req);
 
-      if (scoringQuery.getOriginalQuery() == null) {
-        scoringQuery.setOriginalQuery(context.getQuery());
+      docsWereNotReranked = (rerankingQueries == null || rerankingQueries.length == 0);
+      if (docsWereNotReranked) {
+        rerankingQueries = new LTRScoringQuery[]{null};
       }
-      if (scoringQuery.getFeatureLogger() == null){
-        scoringQuery.setFeatureLogger( SolrQueryRequestContextUtils.getFeatureLogger(req) );
-      }
-      scoringQuery.setRequest(req);
-
-      featureLogger = scoringQuery.getFeatureLogger();
+      modelWeights = new LTRScoringQuery.ModelWeight[rerankingQueries.length];
+      String featureStoreName = SolrQueryRequestContextUtils.getFvStoreName(req);
+      for (int i = 0; i < rerankingQueries.length; i++) {
+        LTRScoringQuery scoringQuery = rerankingQueries[i];
+        if ((scoringQuery == null || !(scoringQuery instanceof OriginalRankingLTRScoringQuery)) && (docsWereNotReranked || (featureStoreName != null && !featureStoreName.equals(scoringQuery.getScoringModel().getFeatureStoreName())))) {
```

Comment:
12/n observations/thoughts/questions:

Most tricky to articulate, hence left until last.

Prior to interleaving the existing logic is that if feature vectors are requested and there is no model (or the model is for a different feature store) then a logging model is created.

So now then if we have two models:
* if both models are for the requested feature store then that's great and each document would have been picked by one of the models and so we use the feature vector already previously calculated by whatever model had picked the document.
* if neither model is for the requested feature store then we need to create a logging model, is one logging model sufficient or do we need two? intuitively to me one would seem to be sufficient but that's based on partial analysis only so far.
* if one of the two models (modelA) is for the requested feature store then for the documents picked by modelA we can use the feature vector already previously calculated by modelA. what about documents picked by modelB? it could be that modelA actually has the feature vector for that document but that modelB simply managed to pick the document first. or if modelA does not have the feature vector then we could calculate it for modelA. would a logging model help in this scenario? intuitively to me it would seem that calculating the missing feature vector via modelA or via the logging model would both be equally efficient and hence no logging model would be needed but again that's only based on partial analysis so far.

Reply:
Spent a lot of time on this, I will continue on Monday, some considerations so far:
- in interleaving the logging model scoring query keeps also the interleaving picks (currently as a subclass of the scoring query), so we would need 2 logging models, by the way, I had various problems due to class cast exceptions and array around, so I will think a bit more about the subclass implementation to understand if it's actually an headache or full benefit, 
- I think the solution will be to implement only one model, 2 different scoringQuery (that contain the picks) and change the feature vector cache to key on: private static int fvCacheKey(LTRScoringQuery scoringQuery, int docid) {
    return  scoringQuery.getScoringModel().hashCode() + (31 * docid);
  }
I keep you posted on monday

Reply:
So I just committed my changes on this, I spent quite a while thinking about the various scenarios, adjusting the code and adding the related tests.

From your observations:
- if both models are for the requested feature store then that's great and each document would have been picked by one of the models and so we use the feature vector already previously calculated by whatever model had picked the document. [OK]
- if neither model is for the requested feature store then we need to create a logging model, is one logging model sufficient or do we need two? intuitively to me one would seem to be sufficient but that's based on partial analysis only so far.
[One is sufficient, and my latest changes do that, anyway I just realized that the loggingModel is not heavy to create and when getting the featureVector, the cache is accessed, the key for that cache doesn't look to the instance but to the content of classes, so two identical logging models would have matched in the feature vector cache, the change was probably not vital, but not harmful either]
- in the third scenario we still need the logging model, because when specifying a featureStore in the featureVector transformer we aim to extract all the features of that store, from efi passed to the transformer, so when explicitly mentioned, we need a logger for both the model again (also for the one that aligns)

Reply:
Actually taking a deeper look to the third point, the original implementation was not extracting all the features, but if the explicit featureStore was matching the model featureStore, it was using the model one (no logger).

So I agree with you, in our case, we want to use the model already existent and no logger at all.
I am going to clean up that bit and do a new commit tomorrow

Reply:
Ok, I have done an extensive refactor of this bit, following your scenarios guideline, I believe the code is much more readable now.
I added few tests as well.
Thank you very much for the insight, now that part is extremely clear.

Before resolving  this discussion, working on this, another consideration sparkled:

Currently when we use the feature logger transformer, all features are extracted (even the ones not used by the reranking model, if any).
Are we sure we want this behavior ?
We could re-use the extracted feature vector  cached also for logging if we just log the features actually used by the model.
I am just wondering why I would be interested in logging for a document, all the features in a featureStore, including potentially features just used by other models.
This could actually lead to confusion.
If you agree I create a separate Jira for that and I'll implement a solution soon, to avoid to forget and context switch.

Reply:
> ... I believe the code is much more readable now ... now that part is extremely clear.

Yes, I agree, very nice.

> ... another consideration sparkled: ... a separate Jira for that ...

Interesting points, will need to think about them a bit (next week). I agree it's unrelated i.e. not a blocker for this pull request here.